### PR TITLE
Make the path that jenkins is accessible on configureable

### PR DESCRIPTION
**Pull request by agentgonzo**
via `--httpPath`
Defaults to `/`

Follow on from #537 allowing for configuration

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: bootstrap/src/main/java/io/jenkins/jenkinsfile/runner/bootstrap/commands/JenkinsLauncherOptions.java [hunk 1]
```diff
@@ -68,6 +68,11 @@
             description = "Port for exposing the web server and Jenkins Web UI from Jenkinsfile Runner. Disabled by default")
     public Integer httpPort;
 
+    @CheckForNull
+    @CommandLine.Option(names = "--httpPath",
+                        description = "Path (including leading /) for exposing the web server and Jenkins Web UI from Jenkinsfile Runner. Root path (/) by default")
```
Comment (jglick) [inline bootstrap/src/main/java/io/jenkins/jenkinsfile/runner/bootstrap/commands/JenkinsLauncherOptions.java]:
```suggestion
                        description = "Path (including leading / but no trailing /) for exposing the web server and Jenkins Web UI from Jenkinsfile Runner. Root path (/) by default")
```

File: setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java [hunk 2]
```diff
@@ -77,7 +77,8 @@ protected ServletContext createWebServer() throws Exception {
                 ? new Server(launcherOptions.httpPort)
                 : new Server(queuedThreadPool);
 
-        WebAppContext context = new WebAppContext(launcherOptions.warDir.getPath(), null);
+        contextPath = launcherOptions.httpPath;
```
Comment (jglick) [inline setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java]: You may need to handle mistakes here:

* trailing `/`
* empty string
* just `/`, which `WebAppContext` will like but `getURL` will not

**Review by jglick — COMMENTED** [review]
Probably YAGNI but :man_shrugging: @oleg-nenashev requested it.

File: setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java (hunk 2 above)
Comment (agentgonzo) [reply setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java]: what do you want me to do with all of those. Just throw an Illegal Argument Exception? I agree that probably YAGNI so I don't want to waste too much time with error cases for something that'll probably not get used

Comment (jglick) [reply setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java]: > Just throw an Illegal Argument Exception?

Sure, just to make things more transparent.

> I don't want to waste too much time with error cases for something that'll probably not get used

Agreed.

File: bootstrap/src/main/java/io/jenkins/jenkinsfile/runner/bootstrap/commands/JenkinsLauncherOptions.java [hunk 3]
```diff
@@ -68,6 +68,11 @@
             description = "Port for exposing the web server and Jenkins Web UI from Jenkinsfile Runner. Disabled by default")
     public Integer httpPort;
 
+    @CheckForNull
+    @CommandLine.Option(names = "--httpPath",
```
Comment (jglick) [inline bootstrap/src/main/java/io/jenkins/jenkinsfile/runner/bootstrap/commands/JenkinsLauncherOptions.java:72]: FYI for `jenkins.war` the analogous option is

```
   --prefix                 = add this prefix to all URLs (eg http://localhost:8080/prefix/resource). Default is none
```

though that name might be confusingly generic-sounding for something that only affects HTTP service which you would normally not use in JFR to begin with.

File: setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java (hunk 2 above)
Comment (agentgonzo) [reply setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java]: done. Case 1 and 3 are covered by the same "ends with /" statement

**Review by jglick — APPROVED** [review]

**Review by fcojfernandez — APPROVED** [review]

File: bootstrap/src/main/java/io/jenkins/jenkinsfile/runner/bootstrap/commands/JenkinsLauncherOptions.java (hunk 3 above)
Comment (oleg-nenashev) [reply bootstrap/src/main/java/io/jenkins/jenkinsfile/runner/bootstrap/commands/JenkinsLauncherOptions.java:72]: I prefer the current option. Maybe `--httpPrefix` but no strong opinion

File: setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java [hunk 4]
```diff
@@ -107,6 +109,15 @@ protected ServletContext createWebServer() throws Exception {
         return context.getServletContext();
     }
 
+    private String validateHttpPath(String httpPath) throws IllegalArgumentException {
+        if (StringUtils.endsWith(httpPath, "/")) {
+            throw new IllegalArgumentException("httpPath must not end with a trailing /");
+        } else if ("".equals(httpPath)) {
```
Comment (oleg-nenashev) [inline setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java:115]:
```suggestion
        } else if (StringUtils.isBlank(httpPath)) {
```

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Documentation would be nice

File: bootstrap/src/main/java/io/jenkins/jenkinsfile/runner/bootstrap/commands/JenkinsLauncherOptions.java (hunk 3 above)
Comment (agentgonzo) [reply bootstrap/src/main/java/io/jenkins/jenkinsfile/runner/bootstrap/commands/JenkinsLauncherOptions.java:72]: I'll leave it as it is then.

File: setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java (hunk 4 above)
Comment (agentgonzo) [reply setup/src/main/java/io/jenkins/jenkinsfile/runner/JenkinsLauncher.java:115]: Not in this case. `null` is a valid input (what you get when the user does not set `--httpPath`) (and `StringUtils.isBlank(null)` returns true)

**Comment by agentgonzo** [comment]
@oleg-nenashev - documentation present now. The build failure seems to be infra-related so can you retrigger a build (or just merge if you are happy as the last commit was only to the readme)

**Review by oleg-nenashev — APPROVED** [review]
Sorry for the review delay. Looks good to me!

**Comment by jglick** [comment]
@oleg-nenashev reminder https://github.com/jenkinsci/jenkinsfile-runner/compare/1.0-beta-29...master

**Pull request merged**